lib/spack/spack/spec.py

```diff
@@ -3511,42 +3507,38 @@ def _dup(self, other, deps=True, cleardeps=True, caches=None):
         self.extra_attributes = other.extra_attributes
         self.namespace = other.namespace
 
-        # Cached fields are results of expensive operations.
-        # If we preserved the original structure, we can copy them
-        # safely. If not, they need to be recomputed.
-        if caches is None:
-            caches = (deps is True or deps == dp.all_deptypes)
-
         # If we copy dependencies, preserve DAG structure in the new spec
         if deps:
             # If caller restricted deptypes to be copied, adjust that here.
             # By default, just copy all deptypes
             deptypes = dp.all_deptypes
             if isinstance(deps, (tuple, list)):
                 deptypes = deps
-            self._dup_deps(other, deptypes, caches)
+            self._dup_deps(other, deptypes)
 
         self._concrete = other._concrete
         self._hashes_final = other._hashes_final
 
-        if caches:
+        if self._concrete:
```

Comment:
The most important aspect of this logical change is that hashes on concrete specs are always preserved (in particular those read from the Cray manifest, which should not recompute their hash).

There might be times when retaining the `caches` parameter could avoid recomputing the hash, but this isn't expected to be a major issue (for one, it shouldn't add any cost when using the ASP concretizer).